pkg/restic/repository_ensurer.go

```diff
@@ -0,0 +1,114 @@
+/*
+Copyright 2018 the Heptio Ark contributors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package restic
+
+import (
+	"context"
+	"sync"
+
+	"github.com/pkg/errors"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/tools/cache"
+
+	arkv1api "github.com/heptio/ark/pkg/apis/ark/v1"
+	arkv1client "github.com/heptio/ark/pkg/generated/clientset/versioned/typed/ark/v1"
+	arkv1informers "github.com/heptio/ark/pkg/generated/informers/externalversions/ark/v1"
+	arkv1listers "github.com/heptio/ark/pkg/generated/listers/ark/v1"
+)
+
+// repositoryEnsurer ensures that Ark restic repositories are created and ready.
+type repositoryEnsurer struct {
+	repoLister arkv1listers.ResticRepositoryLister
+	repoClient arkv1client.ResticRepositoriesGetter
+
+	readyChansLock sync.Mutex
+	readyChans     map[string]chan *arkv1api.ResticRepository
+}
+
+func newRepositoryEnsurer(repoInformer arkv1informers.ResticRepositoryInformer, repoClient arkv1client.ResticRepositoriesGetter) *repositoryEnsurer {
+	r := &repositoryEnsurer{
+		repoLister: repoInformer.Lister(),
+		repoClient: repoClient,
+		readyChans: make(map[string]chan *arkv1api.ResticRepository),
+	}
+
+	repoInformer.Informer().AddEventHandler(
+		cache.ResourceEventHandlerFuncs{
+			UpdateFunc: func(old, upd interface{}) {
+				oldObj := old.(*arkv1api.ResticRepository)
+				newObj := upd.(*arkv1api.ResticRepository)
+
+				if oldObj.Status.Phase != arkv1api.ResticRepositoryPhaseReady && newObj.Status.Phase == arkv1api.ResticRepositoryPhaseReady {
+					r.readyChansLock.Lock()
+					defer r.readyChansLock.Unlock()
+
+					if readyChan, ok := r.readyChans[newObj.Name]; ok {
+						readyChan <- newObj
+						delete(r.readyChans, newObj.Name)
+					}
+				}
+			},
+		},
+	)
+
+	return r
+}
+
+func (r *repositoryEnsurer) EnsureRepo(ctx context.Context, namespace, name string) (*arkv1api.ResticRepository, error) {
+	if repo, err := r.repoLister.ResticRepositories(namespace).Get(name); err != nil && !apierrors.IsNotFound(err) {
+		return nil, errors.WithStack(err)
+	} else if err == nil {
+		if repo.Status.Phase != arkv1api.ResticRepositoryPhaseReady {
+			return nil, errors.New("restic repository is not ready")
+		}
+		return repo, nil
+	}
```

Comment:
Might add a comment here that the rest of the function is for the `apierrors.IsNotFound(err) == true` case.

Reply:
added.